docs: sample GitHub Action for coded app CI/CD (#541)
UiPath/uipath-typescript: docs/coded-apps/ci-cd-github-actions.md

3. **Store the secrets** in the repository settings:
    - Settings → Secrets and variables → Actions → New repository secret.
    - Add `UIPATH_CLIENT_ID` and `UIPATH_CLIENT_SECRET` as **secrets**.

## .github/workflows/deploy.yml

```yaml
name: Deploy Coded App

permissions:
  contents: read

on:
  workflow_dispatch:
    inputs:
      app-name:
        description: "App name to deploy"
        required: true
      path-name:
        description: "URL path name (https://<org>.uipath.host/<path-name>)"
        required: true
      org:
        description: "UiPath organization name"
        required: true
      tenant:
        description: "UiPath tenant name"
        required: true
      folder-key:
        description: "Orchestrator folder key (UUID)"
        required: true

env:
  NODE_VERSION: "20"

jobs:
  deploy:
    runs-on: ubuntu-latest
    steps:
      - name: Checkout
        uses: actions/checkout@v4

      - name: Setup Node.js
        uses: actions/setup-node@v4
        with:
          node-version: ${{ env.NODE_VERSION }}

      - name: Install dependencies
        run: npm ci

      - name: Build
        run: npm run build

>       # Pin to a specific version for reproducible builds

Author: ```suggestion
      # you can also pin to a specific version for reproducible builds
```

      - name: Install codedapp tool
        run: uip tools install codedapp

      # Authenticates using external app client credentials
      - name: Login to UiPath
        run: |
          uip login \
            --organization "${{ inputs.org }}" \
            --tenant "${{ inputs.tenant }}" \
            --client-id "${{ secrets.UIPATH_CLIENT_ID }}" \
            --client-secret "${{ secrets.UIPATH_CLIENT_SECRET }}" \
            --scope "Apps OR.Folders.Read OR.Execution"

      # Creates a .nupkg package from the built app
      - name: Pack
        run: uip codedapp pack ./dist --name "${{ inputs.app-name }}"

      # Uploads .nupkg to Orchestrator and registers the app version
      - name: Publish
        run: uip codedapp publish --name "${{ inputs.app-name }}"

      # Deploys to the specified folder — app becomes available at https://<org>.uipath.host/<path-name>
      - name: Deploy
        run: |
          uip codedapp deploy \
            --name "${{ inputs.app-name }}" \
            --path-name "${{ inputs.path-name }}" \
            --folder-key "${{ inputs.folder-key }}"
```

## Common variations

### Deploy on push to main

> Replace `workflow_dispatch` with a push trigger and use `env` variables instead of `inputs`:

Reviewer: this variation is almost same as the previous one. We could maybe show only the diff here?

# CI/CD: GitHub Actions

A sample GitHub Actions workflow to pack, publish, and deploy a Coded App using the `uip` CLI with external app authentication.

## Prerequisites

Before copying the YAML below:

> 1. **Create a Confidential External Application** in UiPath Admin portal with application scopes: `Apps`, `OR.Folders.Read`, `OR.Administration`, `OR.Execution`.

Reviewer 2: Why is OR.Administrator needed?

Author: not needed, only OR.Execution is needed to upload package.
removed OR.Administration

> 1. **Create an External Application** in UiPath Cloud Portal with scopes: `Apps`, `OR.Folders.Read`, `OR.Administration`, `OR.Execution`.

Author: ```suggestion
1. **Create a Confidential External Application** in UiPath Admin portal with application scopes: `Apps`, `OR.Folders.Read`, `OR.Administration`, `OR.Execution`.
```

> 2. **Assign the external app** to the target Orchestrator folder (Settings → Manage Access → Assign external app).

Author: ```suggestion
2. **Assign the external app** to the Orchestrator folder where you will be deploying your app (Settings → Manage Access → Assign external app).
```

3. **Store the secrets** in the repository settings:
    - Settings → Secrets and variables → Actions → New repository secret.
    - Add `UIPATH_CLIENT_ID` and `UIPATH_CLIENT_SECRET` as **secrets**.

>     - Add `UIPATH_ORG` and `UIPATH_TENANT` as **variables** (not secrets — they're not sensitive).

Author: why are these needed? please use inputs to gh action either

## .github/workflows/deploy.yml

```yaml
name: Deploy Coded App

permissions:
  contents: read

> 

Author: why are we duplicating this at 2 places. one here, and other in samples. just have at one place. i am fine with only having in docs and not having a separate samples fodler.

on:
  workflow_dispatch:
    inputs:
      app-name:
        description: "App name to deploy"
        required: true
      path-name:
        description: "URL path name (https://<org>.uipath.host/<path-name>)"
        required: true
      environment:
        description: "Target environment"
        required: true
        default: "alpha"
        type: choice
        options:
          - alpha
          - staging
          - cloud

env:
  NODE_VERSION: "20"

jobs:
  deploy:
    runs-on: ubuntu-latest
    steps:
      - name: Checkout
        uses: actions/checkout@v4

      - name: Setup Node.js
        uses: actions/setup-node@v4
        with:
          node-version: ${{ env.NODE_VERSION }}

      - name: Install dependencies
        run: npm ci

      - name: Build
        run: npm run build

      - name: Install uip CLI
        run: npm install -g @uipath/uip

      - name: Login to UiPath
        run: |
          uip login \

>             --authority "https://${{ inputs.environment }}.uipath.com" \

Author: remove authority, by default its cloud.uipath.com . we want to document for prod only.

            --organization "${{ vars.UIPATH_ORG }}" \
            --tenant "${{ vars.UIPATH_TENANT }}" \
            --client-id "${{ secrets.UIPATH_CLIENT_ID }}" \
            --client-secret "${{ secrets.UIPATH_CLIENT_SECRET }}" \
            --scope "Apps OR.Folders.Read OR.Administration OR.Execution"

      - name: Pack
        run: uip codedapp pack ./dist

      - name: Publish
        run: uip codedapp publish --name "${{ inputs.app-name }}"

      - name: Deploy
        run: |
          uip codedapp deploy \
            --name "${{ inputs.app-name }}" \

>             --path-name "${{ inputs.path-name }}"

Author: need to pass the folderkey also, and other required flags if any for non interactive

```

> ## Walkthrough

Author: i dont think we need walkthrough sections. if comments needed, add in gh workflow commetns itself

env:
  NODE_VERSION: "20"
  APP_NAME: "my-app"
  PATH_NAME: "my-app"
  ORG: "myorg"
  TENANT: "mytenant"
  FOLDER_KEY: "<your-folder-key>"

jobs:
  deploy:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4
      - uses: actions/setup-node@v4
        with:
          node-version: ${{ env.NODE_VERSION }}
      - run: npm ci
      - run: npm run build
      - run: npm install -g @uipath/cli
      - run: uip tools install codedapp
      - name: Login
        run: |
          uip login \
            --organization "${{ env.ORG }}" \
            --tenant "${{ env.TENANT }}" \
            --client-id "${{ secrets.UIPATH_CLIENT_ID }}" \
            --client-secret "${{ secrets.UIPATH_CLIENT_SECRET }}" \
            --scope "Apps OR.Folders.Read OR.Administration OR.Execution"
      - run: uip codedapp pack ./dist --name "${{ env.APP_NAME }}"
      - run: uip codedapp publish --name "${{ env.APP_NAME }}"
      - name: Deploy
        run: |
          uip codedapp deploy \
            --name "${{ env.APP_NAME }}" \
            --path-name "${{ env.PATH_NAME }}" \
            --folder-key "${{ env.FOLDER_KEY }}"
```

### Pin CLI version

```yaml
- name: Install uip CLI

>   run: npm install -g @uipath/cli@1.4.0

Reviewer 2: Is version needed here?

Author: its not needed, hence sample workflow doesnt have it.
this is part of "Common variation" section, if they want to pin version.
https://github.com/UiPath/uipath-typescript/blob/7bd7ed545723c17ad04e752899c07c87c96cc591/docs/coded-apps/ci-cd-github-actions.md#common-variations

Reviewer 2: Lets remove it. Versions will keep moving. Maybe mention from which version this will work in the doc somewhere.
